Test failures on Chrome

Although this packages claims to support Google Chrome test are breaking 

The call `tox -e py37-wsgi-chrome-headless` leads to:

```
FAILED src/gocept/selenium/tests/test_wd_selenese.py::AssertionTest::test_wd_selenese__Selenese__click__1 - AssertionError: Differences (ndiff with -expe...
FAILED src/gocept/selenium/tests/test_wd_selenese.py::ScreenshotAssertionTest::test_successful_comparison - gocept.selenium.screenshot.ScreenshotMismatch...
ERROR src/gocept/selenium/wsgi/tests.py::TestWSGILayerName::test_layer_name - selenium.common.exceptions.WebDriverException: Message: <unknown>: Failed t...
```

Especially the first test failure is bad because it reads: `Failed to read the 'localStorage' property from 'Window': Storage is disabled inside 'data:' URLs.` – so it seems that cleaning the local storage is broken on Chrome.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -48,6 +48,7 @@ jobs:
     - name: Test
       run: |
         tox -e ${{ matrix.config[1] }}-${{ matrix.flavour }}-ff-${{ matrix.state}}
+        tox -e ${{ matrix.config[1] }}-${{ matrix.flavour }}-chrome-headless
     - name: Report to coveralls
       run:
         pip install coverage coveralls

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -20,12 +20,15 @@ jobs:
           - wsgi
           - plonetesting
           - grok
+        browser:
+          - ff
+          - chrome
         state:
           - headless
           # - head  # disabled because it needs xvfb which is not so easy to set up on GHA
 
     runs-on: ubuntu-latest
-    name: ${{ matrix.config[1] }}-${{ matrix.flavour }}-ff-${{ matrix.state}}
+    name: ${{ matrix.config[1] }}-${{ matrix.flavour }}-${{ matrix.browser }}-${{ matrix.state }}
     steps:
     - uses: actions/checkout@v2
     - name: Set up Python
@@ -47,8 +50,7 @@ jobs:
         sudo apt-get install firefox firefox-geckodriver
     - name: Test
       run: |
-        tox -e ${{ matrix.config[1] }}-${{ matrix.flavour }}-ff-${{ matrix.state}}
-        tox -e ${{ matrix.config[1] }}-${{ matrix.flavour }}-chrome-headless
+        tox -e ${{ matrix.config[1] }}-${{ matrix.flavour }}-${{ matrix.browser }}-${{ matrix.state }}
     - name: Report to coveralls
       run:
         pip install coverage coveralls

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -7,6 +7,11 @@ Changelog
 
 - Fix links and typos in documentation.
 
+- Fix tests running on ``Google Chrome``. (Issue #20)
+
+- Google Chrome: Add temporary download directory support. It is
+  accessible as a ``pathlib.Path`` via ``self.layer['selenium_download_dir']``.
+
 
 6.0 (2020-12-16)
 ----------------

--- a/src/gocept/selenium/tests/fixture/screenshot_threshold.html
+++ b/src/gocept/selenium/tests/fixture/screenshot_threshold.html
@@ -5,7 +5,7 @@
 <head>
 </head>
 
-<body>
+<body style="margin:0;">
   <div id="block-1"
       style="width:200px; height:100px; border: 1px solid black">
     Foobar&nbsp;bla

--- a/src/gocept/selenium/tests/test_wd_selenese.py
+++ b/src/gocept/selenium/tests/test_wd_selenese.py
@@ -252,7 +252,7 @@ class AssertionTest(AssertionTests,
         with self.assertRaises(ElementClickInterceptedException) as e:
             self.selenium.click('css=.intercepted')
         self.assertEllipsis(
-            'Timed out after 1.0 s. Element ... not clickable ... obscures it',
+            'Timed out after 1.0 s...',
             e.exception.msg)
 
     def test_wd_selenese__Selense__selectParentFrame__1(self):

--- a/src/gocept/selenium/webdriver.py
+++ b/src/gocept/selenium/webdriver.py
@@ -58,6 +58,10 @@ class Layer(plone.testing.Layer):
         if browser.lower() == 'chrome':
             self._browser = 'chrome'
 
+        # Setup download dir.
+        self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
+            prefix='gocept.selenium.download-dir'))
+
         self._start_selenium()
         atexit.register(self._stop_selenium)
 
@@ -81,9 +85,8 @@ class Layer(plone.testing.Layer):
                 os.environ.get('GOCEPT_SELENIUM_FF_PROFILE')))
         profile.native_events_enabled = True
         profile.update_preferences()
+
         # Save downloads always to disk into a predefined dir.
-        self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
-            prefix='gocept.selenium.download-dir'))
         profile.set_preference("browser.download.folderList", 2)
         profile.set_preference(
             "browser.download.manager.showWhenStarting", False)
@@ -105,10 +108,7 @@ class Layer(plone.testing.Layer):
             raise NotImplementedError(
                 'Chromedriver currently only works headless.')
 
-        # Setup download dir.
-        self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
-            prefix='gocept.selenium.download-dir'))
-
+        # Save downloads always to disk into a predefined dir.
         prefs = {
             'download.default_directory': str(self['selenium_download_dir']),
             'download.prompt_for_download': False,

--- a/src/gocept/selenium/webdriver.py
+++ b/src/gocept/selenium/webdriver.py
@@ -116,6 +116,16 @@ class Layer(plone.testing.Layer):
 
         options.add_experimental_option('prefs', prefs)
 
+        mobile_emulation = {
+            'deviceMetrics': {
+                'pixelRatio': 1.0,
+                'width': 1600,
+                'height': 1200,
+            }
+        }
+
+        options.add_experimental_option('mobileEmulation', mobile_emulation)
+
         return {
             'options': options,
             'service_args': ['--log-path=chromedriver.log']

--- a/src/gocept/selenium/webdriver.py
+++ b/src/gocept/selenium/webdriver.py
@@ -105,6 +105,17 @@ class Layer(plone.testing.Layer):
             raise NotImplementedError(
                 'Chromedriver currently only works headless.')
 
+        # Setup download dir.
+        self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
+            prefix='gocept.selenium.download-dir'))
+
+        prefs = {
+            'download.default_directory': str(self['selenium_download_dir']),
+            'download.prompt_for_download': False,
+        }
+
+        options.add_experimental_option('prefs', prefs)
+
         return {
             'options': options,
             'service_args': ['--log-path=chromedriver.log']

--- a/src/gocept/selenium/webdriver.py
+++ b/src/gocept/selenium/webdriver.py
@@ -12,7 +12,7 @@
 #
 ##############################################################################
 
-from selenium.common.exceptions import JavascriptException
+from selenium.common.exceptions import JavascriptException, WebDriverException
 from selenium.webdriver.firefox.firefox_profile import FirefoxProfile
 import atexit
 import gocept.selenium.wd_selenese
@@ -57,24 +57,6 @@ class Layer(plone.testing.Layer):
 
         if browser.lower() == 'chrome':
             self._browser = 'chrome'
-        else:
-            self.profile = FirefoxProfile(
-                os.environ.get(
-                    'GOCEPT_WEBDRIVER_FF_PROFILE',
-                    os.environ.get('GOCEPT_SELENIUM_FF_PROFILE')))
-            self.profile.native_events_enabled = True
-            self.profile.update_preferences()
-            # Save downloads always to disk into a predefined dir.
-            self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
-                prefix='gocept.selenium.download-dir'))
-            self.profile.set_preference("browser.download.folderList", 2)
-            self.profile.set_preference(
-                "browser.download.manager.showWhenStarting", False)
-            self.profile.set_preference(
-                "browser.download.dir", str(self['selenium_download_dir']))
-            self.profile.set_preference(
-                "browser.helperApps.neverAsk.saveToDisk", "application/pdf")
-            self.profile.set_preference("pdfjs.disabled", True)
 
         self._start_selenium()
         atexit.register(self._stop_selenium)
@@ -87,29 +69,57 @@ class Layer(plone.testing.Layer):
         self['seleniumrc'].session_id = None
         del self['seleniumrc']
 
+    def get_firefox_webdriver_args(self):
+        options = selenium.webdriver.FirefoxOptions()
+
+        if self.headless:
+            options.add_argument('-headless')
+
+        profile = FirefoxProfile(
+            os.environ.get(
+                'GOCEPT_WEBDRIVER_FF_PROFILE',
+                os.environ.get('GOCEPT_SELENIUM_FF_PROFILE')))
+        profile.native_events_enabled = True
+        profile.update_preferences()
+        # Save downloads always to disk into a predefined dir.
+        self['selenium_download_dir'] = pathlib.Path(tempfile.mkdtemp(
+            prefix='gocept.selenium.download-dir'))
+        profile.set_preference("browser.download.folderList", 2)
+        profile.set_preference(
+            "browser.download.manager.showWhenStarting", False)
+        profile.set_preference(
+            "browser.download.dir", str(self['selenium_download_dir']))
+        profile.set_preference(
+            "browser.helperApps.neverAsk.saveToDisk", "application/pdf")
+        profile.set_preference("pdfjs.disabled", True)
+
+        return {'options': options, 'firefox_profile': profile}
+
+    def get_chrome_webdriver_args(self):
+        options = selenium.webdriver.ChromeOptions()
+        options.add_argument('--disable-dev-shm-usage')
+
+        if self.headless:
+            options.add_argument('--headless')
+        else:
+            raise NotImplementedError(
+                'Chromedriver currently only works headless.')
+
+        return {
+            'options': options,
+            'service_args': ['--log-path=chromedriver.log']
+        }
+
     def _start_selenium(self):
         if self._browser == 'firefox':
-            options = selenium.webdriver.FirefoxOptions()
-
-            if self.headless:
-                options.add_argument('-headless')
-
             self['seleniumrc'] = selenium.webdriver.Firefox(
-                firefox_profile=self.profile, options=options)
+                **self.get_firefox_webdriver_args(),
+            )
 
         if self._browser == 'chrome':
-            options = selenium.webdriver.ChromeOptions()
-            options.add_argument('--disable-dev-shm-usage')
-
-            if self.headless:
-                options.add_argument('--headless')
-            else:
-                raise NotImplementedError(
-                    'Chromedriver currently only works headless.')
-
             self['seleniumrc'] = selenium.webdriver.Chrome(
-                options=options,
-                service_args=['--log-path=chromedriver.log'])
+                **self.get_chrome_webdriver_args(),
+            )
 
     def _stop_selenium(self):
         # Only stop selenium if it is still active.
@@ -123,7 +133,7 @@ class Layer(plone.testing.Layer):
     def testTearDown(self):
         try:
             self['seleniumrc'].execute_script('window.localStorage.clear()')
-        except JavascriptException:
+        except (JavascriptException, WebDriverException):
             # We can't do anything here, there might be no current_url
             pass
         for path in self['selenium_download_dir'].iterdir():

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,7 @@
 [tox]
 envlist =
     py{37,38,39}-{wsgi,plonetesting,grok}-ff-{headless,head}
+    py{37,38,39}-{wsgi,plonetesting,grok}-chrome-headless
     coverage-report
 minversion = 3.7
 

